### ISSUE TITLE
Wine WOW builds don't support 32-bit gstreamer 

### ISSUE THREAD

**Issue by kurnevsky** [NixOS/nixpkgs#132352]
### Describe the bug
It seems wine WOW builds don't support 32-bit gstreamer. There are such messages in wine [build](https://hydra.nixos.org/build/149240907) [logs](https://nix-cache.s3.amazonaws.com/log/j2agrpk7rkrssw1rcyxscy8dh2q1kg05-wine-wow-6.0.drv):

```
configure: libhal 32-bit development files not found, no legacy dynamic device support.
configure: libusb-1.0 32-bit development files not found (or too old), USB devices won't be supported.
configure: gstreamer-1.0 base plugins 32-bit development files not found, GStreamer won't be supported.
configure: OSS sound system found but too old (OSSv4 needed), OSS won't be supported.
configure: libcapi20 32-bit development files not found, ISDN won't be supported.
configure: libcups 32-bit development files not found, CUPS won't be supported.
configure: libkrb5 32-bit development files not found (or too old), Kerberos won't be supported.
configure: vkd3d 32-bit development files not found (or too old), Direct3D 12 won't be supported.
```
Not sure about other libs in this list, but missing gstreamer causes errors when games try to play mp3 files.

### Steps To Reproduce
For instance, try to run Morrowind with wine32-full and wineWow-full. With the latter it will have an error in the main menu that says that it can't play mp3 file.

### Expected behavior
Wine WOW is able to run Morrowind.

### Notify maintainers
@avnik @7c6f434c @bendlas

### Metadata
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.13.5-zen1, NixOS, 21.05.20210731.11c6620 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210707_02dd6bb`
 - channels(root): `"nixos-21.05.1970.11c662074e2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by CobaltCause** [comment]
This is still relevant. In particular, I discovered this because EA's Origin crashes after running for a while due to errors like `(wine:1436525): GStreamer-WARNING **: 18:13:19.092: Failed to load plugin [...] Wrong ELF class: ELFCLASS64`.

**Comment by SuperSamus** [comment]
On the wiki page there is this, I don't know if it's relevant:
> ⚠ Glib2 and gstreamer have different header files for 32 and 64 bit, so when compiling 32 bit Wine with gstreamer support on a 64 bit system it is necessary to use PKG_CONFIG_PATH=/path/to/pkgconfig to point to the location of the 32 bit files when running ./configure. Without that, ./configure will find the 64 bit development files instead of the 32 bit ones and disable gstreamer support. (See https://www.winehq.org/pipermail/wine-devel/2016-January/111245.html)

**Comment by bendlas** [comment]
May have a point there. Right now, we're adding both 32 and 64 bit libs into buildInputs: https://github.com/NixOS/nixpkgs/blob/f71c25b983d2724a3e98266ebee801088f504856/pkgs/applications/emulators/wine/base.nix#L66

**Comment by SuperSamus** [comment]
I wonder if it's related to #207641...

Speaking of that bug (which is about the fact that binary Wine downloads are unable to use 32 bit GStreamer, when using FHS environments like Bottles)...
Some 32 bit games, when I play them in an FHS environment like Bottles, do have a working GStreamer, if I use system Wine WOW (from nixpkgs). I experience this for example with "Ultimate Spider-Man" or "Spider-Man: Shattered Dimensions".

I play them without Bottles? Crash.
I play them with Bottles? Working videos.

But not all games do this (e.g. Oblivion, which has not working audio in both cases).

(Note: this comment may be a false alarm that deserves a separate issue, since Wine on my end seems to have problems even without GStreamer, if I use it outside a FHS environment. For example, with 32 bit Wine, Ultimate Spider-Man crashes on boot anyway, while Spider-Man Shattered Dimensions has huge graphical issues)

**Comment by jys1670** [comment]
Still relevant. It happened again (on latest unstable) with specific game on `wineWowPackages.unstableFull` package; however the very same game was working fine on wine 8.5, somewhere within a week timeframe before [this update](https://github.com/NixOS/nixpkgs/pull/237206). I do not claim that the update caused a regression, most likely issue is about changes in dependent packages.

**Comment by jys1670** [comment]
Bisected to [7c37217a76ffc97c7dc5535d919f22dcc71f0d9b](https://github.com/NixOS/nixpkgs/pull/237206/commits/7c37217a76ffc97c7dc5535d919f22dcc71f0d9b)
with error 
```
ERROR: Caught a segmentation fault while loading plugin file:
/nix/store/24plnj748vqm2a3v7wvhsj8h6fsvabsc-gst-plugins-bad-1.22.3/lib/gstreamer-1.0/libgstva.so
```
@SuperSamus have you tried using wineWow 8.10 with any 32-bit game? If so, do you have any errors like that one? It's weird.

**Comment by SuperSamus** [comment]
What game exactly?
I tried running Sonic Mania (which is 32 bit) on a fresh prefix, and it runs fine.

Also, the error:
> GStreamer-WARNING ... : Failed to load plugin ... wrong ELF class: ELFCLASS64
Doesn't necessarily mean that something is wrong, it's possible that it tries to load the 64 bit plugin first, and then successfully loads the 32 bit plugin with no log.

Have you tried on a fresh prefix?

**Comment by jys1670** [comment]
Game is Cyanotype Daydeam. Sadly fresh prefix does not fix it.
What bothers me is that Wine 8.05 does not have these warnings, yet 8.10 does. Nix is supposed to be all about reproducibility, so different wine versions looking for libraries in different order in weird. Especially when they have identical wrappers (literally compared them with diff, both wines point to the same gstreamer related files in the same order, then launch `.wine` of corresponding version)
Strace logs are also similar. I took a quick look: before crash there was load of  `~/.cache/gstreamer-1.0/registry.i686.bin`, removing it didn't fix anything. There might be other environment impurities that could potentially break things.

**Comment by SuperSamus** [comment]
Shot in the dark, but maybe this is a regression upstream?

**Comment by jys1670** [comment]
> Shot in the dark, but maybe this is a regression upstream?

That's unlikely, otherwise we would have seen a lot of similar reports.

Turns out root of all evil is nvidia (haha, classics). I have optimus laptop / hybrid graphics (nvidia prime), yet game has never been launched on dGPU, iGPU was used instead. But for the sake of experimenting I decided to completely remove nvidia from system config and did nixos-rebuild after that. Same warnings were printed, but game worked in the end!

So nvidia randomly breaks things on wineWow 8.10 and works on wineWow 8.05, no idea why.
At the same time `nix-shell -p winePackages.unstable` (pure 32-bit only wine 8.10) works fine on nvidia + intel setup.

**Comment by SuperSamus** [comment]
Let's go back to the original issue, though.
The problem is that pkgconfig is bad at picking the right architecture. (As I said on https://github.com/NixOS/nixpkgs/issues/132352#issuecomment-1454751514)

I see two ways to fix this:
1) The dirty way is in `builder-wow.sh`, before starting to configure for the 32 build, to get `PKG_CONFIG_PATH` and prune out all 64 bit derivations. Problem is... how do I do that?

2) Rework how `wine-wow` is built. Instead of building everything in a single derivation, split `wine64` in two derivations: one that stops after building (let's call it `wine64Build`), and the other (the actual `wine64`) that finishes from where wine64Build left off. This way, we can use `wine64Build` as a starting point for `wine-wow`, and have only 32 bit dependencies in the environment.
Problem is, in nixpkgs Wine wasn't set up for doing all this. For example, staging patches are obtained by overriding the derivation, meaning that `wine64Build` won't be patched since it's just an input. Other example, Wine uses a fully fledged builder that isn't meant to be divided in phases like this.

**Comment by bendlas** [comment]
> from https://github.com/NixOS/nixpkgs/pull/238991#issue-1767637334
> This is so dirty...

oof ok. I think longer term, exporing your second proposal of re-using the 64bit build's build directory could be well worth it.

Originally, I created `builder-wow.sh` as a straight port for https://wiki.winehq.org/Building_Wine#Shared_WoW64 but that's been a few years back and may have been updated / superseded, so it may be worth revisiting that at some point

**Comment by SuperSamus** [comment]
On the other hand, though, upstream is working to remove the need for 32 bit dependencies in the first place. When completed, that would make WoW obsolete and make this effort a bit useless.
(Info on https://www.winehq.org/announce/8.0.)
